packages/block-editor/src/components/inserter/media-tab/hooks.js

```diff
@@ -1,89 +1,183 @@
+ * @property {string}        [alt]        The alt text of the media item.
+ * @property {string}        [caption]    The caption of the media item.
+ *
+ */
+
+/**
+ * Fetches media items based on the provided category.
+ * Each media category is responsible for providing a `fetch` function.
+ *
+ * @param {Object}               category The media category to fetch results for.
+ * @param {InserterMediaRequest} query    The query args to use for the request.
+ * @return {InserterMediaItem[]} The media results.
+ */
+export function useMediaResults( category, query = {} ) {
+	const [ mediaList, setMediaList ] = useState();
+	const [ isLoading, setIsLoading ] = useState( false );
+	// We need to keep track of the last request made because
+	// multiple request can be fired without knowing the order
+	// of resolution, and we need to ensure we are showing
+	// the results of the last request.
+	// In the future we could use AbortController to cancel previous
+	// requests, but we don't for now as it involves adding support
+	// for this to `core-data` package.
+	const lastRequest = useRef();
 	useEffect( () => {
 		( async () => {
-			setResults();
-			const _media = await settings?.__unstableFetchMedia( options );
-			if ( _media ) setResults( _media );
+			const key = JSON.stringify( {
+				category: category.name,
+				...query,
+			} );
+			lastRequest.current = key;
+			setIsLoading( true );
+			setMediaList( [] ); // Empty the previous results.
+			const _media = await category.fetch?.( query );
+			if ( key === lastRequest.current ) {
+				setMediaList( _media );
+				setIsLoading( false );
+			}
 		} )();
-	}, Object.values( options ) );
-	return results;
+	}, [ category.name, ...Object.values( query ) ] );
+	return { mediaList, isLoading };
+}
+
+function useInserterMediaCategories() {
+	const { inserterMediaCategories, allowedMimeTypes } = useSelect(
+		( select ) => {
+			const { getSettings } = select( blockEditorStore );
+			const {
+				__unstableInserterMediaCategories,
+				allowedMimeTypes: _allowedMimeTypes,
+			} = getSettings();
+			return {
+				inserterMediaCategories: __unstableInserterMediaCategories,
+				allowedMimeTypes: _allowedMimeTypes,
+			};
+		},
+		[]
+	);
+	// The allowed `mime_types` can be altered by `upload_mimes` filter and restrict
+	// some of them. In this case we shouldn't add the category to the available media
+	// categories list in the inserter.
+	const allowedCategories = useMemo( () => {
+		if ( ! inserterMediaCategories || ! allowedMimeTypes ) {
+			return;
+		}
+		return inserterMediaCategories.filter( ( category ) => {
+			// When a category has set `hasAvailableMedia` to `true`, means is an
+			// external source. In this case we don't need to check for allowed mime types,
+			// as they are used for restricting uploads for this media type and not for
+			// inserting media from external sources.
+			if ( category.hasAvailableMedia ) {
```

Comment:
In that case, why `hasAvailableMedia` and not a possibly clearer `isExternalResource`?